Source/Graphic/HCRaster+File.c

```diff
@@ -0,0 +1,348 @@
+//
+//  HCRaster+File.c
+//  HollowCore
+//
+//  Created by Matt Stoker on 4/18/20.
+//  Copyright © 2020 HollowCore. All rights reserved.
+//
+
+#include "HCRaster+File.h"
+
+//----------------------------------------------------------------------------------------------------------------------------------
+// MARK: - File Operations
+//----------------------------------------------------------------------------------------------------------------------------------
+HCRasterRef HCRasterCreateByLoadingPPM(const char* path) {
+    // Open a file for reading
+    FILE* f = fopen(path, "r");
+    if (f == NULL)
+        return NULL;
+    
+    // Verify file magic number
+    int magic0 = fgetc(f);
+    int magic1 = fgetc(f);
+    if (magic0 != 'P' || (magic1 != '3' && magic1 != '6')) {
+        return NULL;
+    }
+    
+    // Read and validate the size and channel scale from the header and consume the newline sequence
+    int width = 0;
+    int height = 0;
+    int channelMax = 0;
+    if (fscanf(f, "%i %i %i", &width, &height, &channelMax) != 3 || width <= 0 || height <= 0 || channelMax <= 0 || channelMax > 255) {
+        return NULL;
+    }
+    if (fgetc(f) == '\r') {
+        char c = fgetc(f);
+        if (c != '\n') {
+            ungetc(c, f);
+        }
+    }
+    HCBoolean binary = magic1 == '6';
+    
+    // Trash any comment lines
+    if (!binary) {
+        while (true) {
+            // Read a comment start character or return it to the stream and continue
+            int c = fgetc(f);
+            if (c != '#') {
+                ungetc(c, f);
+                break;
+            }
+            
+            // Read until a newline, accounting for possible \r\n sequences
+            while (true) {
+                c = fgetc(f);
+                if (c == '\n' || c == EOF) {
+                    break;
+                }
+                if (c == '\r') {
+                    c = fgetc(f);
+                    if (c != '\n') {
+                        ungetc(c, f);
+                    }
+                    break;
+                }
+            }
+        }
+    }
+    
+    // Allocate raster
+    HCRasterRef self = HCRasterCreate(width, height);
+    if (self == NULL) {
+        return NULL;
+    }
+    
+    // Read pixel information
+    for (HCInteger yIndex = 0; yIndex < HCRasterHeight(self); yIndex++) {
+        for (HCInteger xIndex = 0; xIndex < HCRasterWidth(self); xIndex++) {
+            int r = 0;
+            int g = 0;
+            int b = 0;
+            if (!binary) {
+                fscanf(f, "%i %i %i", &r, &g, &b);
```

Comment:
Should we ensure we are reading 3 values here? If it fails that means we have an invalid file and should probably fail.

Reply:
Now fails.